feat: Implement a new orthogonal range search seed finder

stephenswat opened the pull request:
As I said in #901, I have been playing around with seed finding a little bit lately. Last weekend, I mentioned an idea for a new (?) kind of seed finding algorithm based on range search datastructures, and this is the very, very first semi-working implementation of it, just before the weekend.

The idea behind this algorithm is relatively simple. In traditional seedfinding, we check a whole lot of candidate spacepoints to see whether they meet some condition. If you look at this differently, each spacepoint defines a volume in the z-r-φ space, which contains any spacepoints it can form a doublet with. What if we reversed this logic? What if we defined this volume first, and then just extract the spacepoints inside of that space? That way, we can vastly reduce the number of spacepoints we need to look at.

How do we do this quickly? With [_k_-d trees](https://en.wikipedia.org/wiki/K-d_tree). These data structures are cheap to build, and they give us very fast orthogonal range searches. In other words, we can very quickly look up which of our spacepoints lie within an axis-aligned orthognal n-dimensional hyperrectangle. In this case, which spacepoints lie within a z-r-φ box.

So, the core idea of this seedfinder is to define as many of our seedfinding constraints in orthogonal fashion. That way, we can make our candidate hyperrectangle smaller and smaller. The tighter the constraints we can place, the better. Then, we look up the relevant spacepoints, and we can avoid looking at any others. That also means this solution requires no binning whatsoever.

## Constraint conversion

Currently there are quite a few constraints in the code. Here is my status update on how well it is going to convert each of them. In some cases, we can define a weaker version of the constraints in orthogonal fashion. This is still very powerful, and it doesn't actually lose us any efficiency (because we can always check the tighter constraint in a non-orthogonal way later, not a problem)!

### Unary constraints

Currently, I am not aware of any unary constraints in the Acts seed finding code. That is to say, logic to determine whether a point is allowed to be a lower spacepoint. However, I have the following thoughts about introducing some:

* I believe the binning code does some kind of magic to determine whether a spacepoint can be a lower spacepoint. Since my solution doesn't use any binning, I don't have access to this just yet. However, if we can incorporate this logic it could be very powerful.
* Maximum single-point η: we currently have some checks in place to see if the pseudorapidity of particles is not too high. We could realistically use this maximum pseudorapidity, combined with the collision region range to constrain the bottom spacepoints.

### Binary constraints

These are the existing binary constraints on spacepoint duplets:

Constraint | Description | Orthogonalization
-|-|-
Minimum ∆r | Ensure that the second spacepoint is within a certain difference in radius | Full
Maximum cot θ | Ensure that the pseudorapidity of the duplet is not too high | Unsuccessful
z-origin range | Ensure that the duplet would have originated from the collision point | Weakened 
Maximum ∆φ<sup>1</sup> | Ensure that the duplet does not bend too much in the x-y plane | Full

<sup>1</sup> This check does not exist explicitly in the existing seed finder, but is implicit in the binning process.

### Ternary constraints

There are a lot of ternary constraints (to check whether a triplet is valid):

Constraint | Description | Orthogonalization
-|-|-
Scattering angle | ??? | Unsuccessful
Helix diameter | Ensure the helix diameter is within some range | In progress
Impact parameters | Ensure the impact parameters are close to the collision point | In progress
Monotonic z<sup>1</sup> | Ensure that z increases or decreases monotonically between points | Full

<sup>1</sup> This check does not exist in the existing seed finder, check #901.

There are also constraints defined in the experiment-specific cuts, and the seed filter, and in other places. If we could convert some of those to orthogonal constraints the implementation would become much more powerful. However, I don't really understand what is happening in those files just yet. Need more reading.

## Current performance

The current performance of this seedfinder is... Complicated. On my machine, it runs a 4000 π<sup>+</sup> event in about 5 seconds, three times slower than the existing seedfinder. Its efficiency is much higher though, and the fake rate is much lower. So that's something. However, that is in part because I am creating far more seed candidates, so take this with a big grain of salt.

## Potential gain

There are two ways that I can think of to use this kind of algorithm. The first is an inside-to-outside algorithm, where we pick a lower spacepoint first, check the space it defines for a middle spacepoint, and then check the space the two of them define for a third spacepoint. This algorithm has time complexity 𝒪(_n_<sup>3</sup>), and it has space complexity 𝒪(_n_). Due to the constants, I still believe this implementation can outperform the 𝒪(_n_<sup>2</sup>) existing algorithm, however.

The second way would be to construct a set of duplets using this logic, and then to fit those together like we do with traditional seedfinding. This has 𝒪(_n_<sup>2</sup>) time complexity like the existing code, but also space complexity 𝒪(_n_<sup>2</sup>).

## Things that are completed

* The implementation of the _k_-d tree seems to work very well, and it is quite fast.
* Basic seedfinding using this strategy is functional.

## Things that don't work yet

* My maximum ∆φ constraint does not cross the 2π boundary yet.
* I used the existing seedfinding algorithm as a stepping stone, which I have completely destroyed in the process. Obviously I do not intend on keeping it that way, and the existing algorithm will be restored to its full glory.
* Lots more.

## Things that can be improved

* Add more constraints, and tighten existing ones.
* Lots of things, pretty much everything. But I really want to go home for the weekend, so I will write this part next week.

asalzburger commented:
@stephenswat I will try to have a look over it, but I can not promise this to happen before next week.

stephenswat commented:
As a little update, I have worked on the performance (in terms of physics and compute), and they have improved a lot since I opened this PR. Here are some extremely preliminary results between the implementations:

### Existing seedfinder
```
SeedingPerfo   DEBUG     nTotalSeeds               = 21032
SeedingPerfo   DEBUG     nTotalMatchedSeeds        = 6269
SeedingPerfo   DEBUG     nTotalParticles           = 3765
SeedingPerfo   DEBUG     nTotalMatchedParticles    = 2791
SeedingPerfo   DEBUG     nTotalDuplicatedParticles = 1543
SeedingPerfo   INFO      Efficiency (nMatchedParticles / nAllParticles) = 0.741301 (2791 / 3765)
SeedingPerfo   INFO      Fake rate (nUnMatchedSeeds / nAllSeeds) = 0.70193 (14763 / 21032)
SeedingPerfo   INFO      Duplication rate (nDuplicatedMatchedParticles / nMatchedParticles) = 0.552848
SeedingPerfo   INFO      Average number of duplicated seeds ((nMatchedSeeds - nMatchedParticles) / nMatchedParticles) = 1.24615
Sequencer      DEBUG       Algorithm:SeedingAlgorithm: 626.734145 ms/event
```

### My orthogonal seedfinder
```
SeedingPerfo   DEBUG     nTotalSeeds               = 18496
SeedingPerfo   DEBUG     nTotalMatchedSeeds        = 8169
SeedingPerfo   DEBUG     nTotalParticles           = 3765
SeedingPerfo   DEBUG     nTotalMatchedParticles    = 3354
SeedingPerfo   DEBUG     nTotalDuplicatedParticles = 2287
SeedingPerfo   INFO      Efficiency (nMatchedParticles / nAllParticles) = 0.890837 (3354 / 3765)
SeedingPerfo   INFO      Fake rate (nUnMatchedSeeds / nAllSeeds) = 0.558337 (10327 / 18496)
SeedingPerfo   INFO      Duplication rate (nDuplicatedMatchedParticles / nMatchedParticles) = 0.681872
SeedingPerfo   INFO      Average number of duplicated seeds ((nMatchedSeeds - nMatchedParticles) / nMatchedParticles) = 1.4356
Sequencer      DEBUG       Algorithm:SeedingAlgorithm: 720.834925 ms/event
```

This is with the same 4000 pion event. Of course, these results are _very_ preliminary, and there could be all sorts of problems hiding here.

stephenswat commented:
I've tightened the constraints to be more similar to the constraints imposed by the binning in the existing implementation. This has given me some nice performance results:

```
SeedingPerfo   DEBUG     nTotalSeeds               = 13148
SeedingPerfo   DEBUG     nTotalMatchedSeeds        = 6463
SeedingPerfo   DEBUG     nTotalParticles           = 3765
SeedingPerfo   DEBUG     nTotalMatchedParticles    = 3296
SeedingPerfo   DEBUG     nTotalDuplicatedParticles = 1626
SeedingPerfo   INFO      Efficiency (nMatchedParticles / nAllParticles) = 0.875432 (3296 / 3765)
SeedingPerfo   INFO      Fake rate (nUnMatchedSeeds / nAllSeeds) = 0.508442 (6685 / 13148)
SeedingPerfo   INFO      Duplication rate (nDuplicatedMatchedParticles / nMatchedParticles) = 0.493325
SeedingPerfo   INFO      Average number of duplicated seeds ((nMatchedSeeds - nMatchedParticles) / nMatchedParticles) = 0.960862
Sequencer      DEBUG       Algorithm:SeedingAlgorithm: 192.789488 ms/event
```

That makes this implementation roughly three times faster then the existing one. However, as before, I am not ready to make any claims about the quality of these seeds, so please be wary!

I've also implemented a secondary seeding strategy. The main seeding strategy here is similar to what we are already doing: find a middle spacepoint, then for that middle spacepoint find bottom and top candidates, and then explore the cartesian product of those sets. My new strategy starts with a middle spacepoint, then finds all the candidate bottom spacepoints, and then uses constraints imposed by both the middle and bottom spacepoint to tightly constrain the top spacepoint. However, this strategy is slower than the other strategy, for now.

stephenswat commented:
I've made some further strides in terms of performance:

```
Sequencer      DEBUG       Algorithm:SeedingAlgorithm: 62.982515 ms/event
```

However, this pull request is also becoming rather large. Perhaps when the time comes I can split it up into two: one to introduce the k-d tree code, and one to introduce the seed finder based on it.

stephenswat commented:
Some efficiency plots. Please take with a large grain of salt because while I have tried to mimic the configuration of the existing implementation as much as possible, it's possible I may have overfitted significantly. Tested on data from Fatras with flags `--event=100 --bf-constant-tesla=0:0:2 --gen-nparticles=4000 --gen-pdg 211`.

![eff_eta](https://user-images.githubusercontent.com/1165820/128935301-dc27b68c-7771-4479-87ca-8d5b17fcf1ca.png)
![eff_phi](https://user-images.githubusercontent.com/1165820/128935303-29c30030-13e2-4030-849e-706f9dfa8ce6.png)
![eff_pT](https://user-images.githubusercontent.com/1165820/128935304-2ce927a8-008e-42c4-9e4a-a553cb9ee847.png)

stephenswat commented:
Here is an interesting geometrical construction that allows us to constrain the position of the third spacepoint according to the minimum helix diameter and the maximum impact parameter:

1. Fix points _A_ and _B_ in the x-y plane, with _A_ closest to the origin.
1. Add a circle _I_ with the minimum impact parameter radius around the origin.
1. Draw a line _e_ between the origin and _A_.
1. Define a point _D_ as the intersection of _I_ and _e_.
1. Define a point _E_ halfway between _A_ and _D_.
1. Add a hyperbola _f_ with loci _A_ and _O_, passing through _E_.
1. Draw the perpendicular bisector _g_ of _A_ and _B_.
1. Define points _F_ and _G_ as the intersections of bisector _g_ and hyperbola _f_.
1. Draw two circles, _M_ and _N_, with centers _F_ and _G_ respectively, both passing through A. Note that these two circles are now tangent circles to the circle _I_, our minimum impact parameter.
1. Draw a circle _Q_ with center _A_ and a radius equal to the minimum helix radius.
1. Find the intersections _H_ and _J_ between the circle _Q_ and the perpendicular bisector _g_.
1. Draw circles _U_ and _V_ through _H_ and _J_, both with radius equal to the minimum helix radius.

![Screenshot from 2021-08-11 01-03-31](https://user-images.githubusercontent.com/1165820/128946136-b4631e35-4659-4cb9-9c02-6f736a65618f.png)

The third spacepoint _C_ is now constrained: it must lie outside of _M_ and _N_ (the red circles in the diagram above), otherwise its perigee is too far from the origin. It must also lie outside of _U_ and _V_ (the blue circles), because if they were inside either of those circles the helix radius would be too small to fit the constraints.

stephenswat commented:
And some compute performance results. Again, I am not trying to make any claims about this implementation, I am almost certain there is something I have missed or done wrong. Just want to share some early results. :smile:

![perf](https://user-images.githubusercontent.com/1165820/128947793-7e9271f1-71bc-4e14-ae4d-eead6211beb0.png)

And with a logarithmic y-axis for readability:

![perf_zoom](https://user-images.githubusercontent.com/1165820/128947795-d23b63be-ebaf-4034-8440-902857d2b3a0.png)

paulgessinger commented:
Can you try this with nonzero `gen-vertex-xy-std-mm` and `gen-vertex-z-std-mm` if you haven't already?

stephenswat commented:
Do you have any suggestions for sensible values to use for these flags?

paulgessinger commented:
Hm, so ATLAS luminous region data is here: https://twiki.cern.ch/twiki/bin/view/AtlasPublic/BeamSpotPublicResults
But I think we only have these single stddev parameters to tune, while for ATLAS the size plays a role, but also where the beam actually is. I guess we can approximate by taking the larger of the two (basically, our gaussian covers both the beam spot size and position variation).

Based on this, I'd say something like 1-2mm in xy and 50-100mm in z.

paulgessinger commented:
It might also be interesting to look at the seed purity, or fake rate as a figure of interest.

robertlangenberg commented:
Very nice results, this looks super promising! Do I understand correctly that you (so far) "only" replace the binning and use the logic of the seedfinder as is? If that's the case, and you'd like to have a baseline for what the best efficiency could be, you can run the seeding without binning. That's going to take a while, not sure how long, so I suggest to start with 1 event only ;)

I'd be very happy to discuss configuration parameters in detail or any of your improvements and how I'd expect them to compare to the seeding implementation.

stephenswat commented:
Thanks for your comment @robertlangenberg! There are actually two strategies here, one is very similar to the existing seed finding, only the way candidate duplets are found is replaced by the k-d tree search method. The second strategy is a little different, which take a middle spacepoint, iterates over the bottom spacepoints, and then searches in a space constrained by both the bottom and the middle spacepoint for a top spacepoint. So that strategy is a little different.

Regarding the efficiency problems we talked about - I ran some tests. You may want to check my logic on this, but my findings are as follows. Using the standard configuration defined in `SeedingExample.cpp` (which we agreed was not really optimized for the "generic" detector), all three implementations (the existing one and my two new ones) drop dramatically in efficiency to about 30%. However, if I increase the `maxSeedsPerSpM` parameter from 1 to 100, the efficiency recovers to 99% both all implementations. Of course this massively increases the fake rate, but it indicates to me that there is no problem in the initial seed finding step, but that the good seeds are being thrown out in the seed filtering step. In other words, I don't think there is any problem in the actual code, just in the configuration: the configuration is too loosely constrained, creating too many bogus seeds, and thereby the real seeds are being pushed out of the queue by the seed filtering. I hope I have interpreted that correctly. :smile:

beomki-yeo commented:
> just in the configuration: the configuration is too loosely constrained, creating too many bogus seeds, and thereby the real seeds are being pushed out of the queue by the seed filtering

I think stanford group worked (possibly last year?) on the seedfinding optimization by tuning the config parameters. I don't think they implemented their optimization into main repository yet. @robertlangenberg do you know any source about it?

paulgessinger commented:
@beomki-yeo I think it's here https://github.com/Pchatain/seedingWithEA

I've been looking into this, and plan to push this towards integration in core.

stephenswat commented:
So, a little update. I've put this project on the backburner a little the past week or so, and we had some setbacks. Paul's suggestion of trying a variable beam spot exposed a lot of problems in my approach, and when I got them fixed most of the speedup was gone. Since then I have tried to get that speedup back, and I have made some advances there. I have added some additional orthogonal cuts to the seeds (notably some of these could also be used in the binning-based seed finder, so keep that in mind when evaluating these results). Here are some plots to show the current state. As before, I tested the orthogonal approach versus the existing binning approach at event sizes ranging from 1 to 12000 pions, each time testing 100 events. All of this data was generated using ATLAS-like beam spot deviation, and the configuration as it is in the repository right now.

### Runtime

Runtime is of course the primary motivator for this project. Currently, the runtime looks promising:

![runtime](https://user-images.githubusercontent.com/1165820/130773527-dfc499ba-4ef4-4023-bc28-759b5b50aa52.png)
![speedup](https://user-images.githubusercontent.com/1165820/130773525-cf2cd508-6b5b-4244-bdfe-9345354eed41.png)

At HL-LHC levels of 4000 pions, my orthogonal approach is about six times faster than the binning-based approach. As the pile-up increases, so does the speedup factor.

### Quality

Of course we must also check the seed quality, and my original approach fell down here when we had non-zero primary vertex positions. I am happy to report that this has improved a lot, although there are some areas where I still lag behind the binning-based seed finder.

![efficiency](https://user-images.githubusercontent.com/1165820/130773522-6ca9503b-5fa5-45a5-82b1-60ae9b715dc1.png)

In terms of efficiency, at the HL-LHC parameters my orthogonal approach is about 0.3% less efficient than the binning approach. This isn't great, and I aim to get this up. I am quite confident that this will be possible through some minor modifications of the new cuts that I have added.

![fakerate](https://user-images.githubusercontent.com/1165820/130773523-c157bfdd-ca95-4b97-9521-02592887845f.png)

The track purity shows some interesting phenomena. Up to 5000 pions, my orthogonal approach has better track purity than the binning approach. At 5500 this switches around, and from 6000 pions and higher the binning approach actually has slightly higher purity (lower fake rate) than my approach.

### Efficiency at 4000 _π<sup>+</sup>_

Here is a more in-depth analysis of the efficiency around the HL-LHC mark. As shown earlier, my approach has slightly lower efficiency, specifically in three regimes. Firstly, there is reduced efficiency at low pseudorapidities. Secondly, there is slighly decreased efficiency around the edges of the phi ranges. This is most likely because my approach does not yet support phi wrapping. Lastly, I see some minor drop in performance at high _p<sub>T</sub>_, which I must admit I don't really understand.

![eff_pT](https://user-images.githubusercontent.com/1165820/130773516-131c82d0-16d7-4085-8d3f-8f91b07023e9.png)
![eff_eta](https://user-images.githubusercontent.com/1165820/130773518-b7ef3cb8-762a-47e7-9430-eb1640874b11.png)
![eff_phi](https://user-images.githubusercontent.com/1165820/130773520-e389dae2-a7b3-470c-98ad-a3a4a0bdb3dc.png)

asalzburger commented:
Hi @stephenswat - this is really impressive work.

I have discussed with @noemina that we should try to get the standard seed finder (+ the ITk updated version) as well as this new alternative implementation to be tested on the ITK.

Hence, I would be super happy if we could progress on moving that towards PR merging.

asalzburger commented on Core/include/Acts/Seeding/SeedfinderUtils.hpp:
```diff
@@ -0,0 +1,240 @@
+#pragma once
```
The file will need an ACTS header.

Also, it will have to apply with ACTS Coding standards, which is CamelCase.

asalzburger commented on Core/include/Acts/Seeding/SeedfinderUtils.hpp:
```diff
@@ -0,0 +1,240 @@
+  return _lincircle(middle, top, false);
+}
+
+template<typename SP>
```
Where do these numbers come? Is this from the original Seeding code - I think that's super tailored to ATLAS.

@robertlangenberg - can you comment on that?

asalzburger commented on Core/include/Acts/Seeding/SeedfinderUtils.hpp:
```diff
@@ -0,0 +1,240 @@
+template<typename SP>
+bool validMiddle(
+  const SP & s,
+  const Acts::SeedfinderConfig<ActsExamples::SimSpacePoint> & config
```
That seems to be mixing `ActsExamples` definition within `Core` - that's a no go.

What I think we should do is the following:
- The seed finder can have some `struct` that allows you to call `maxDeltaR` and/or `validMiddle` etc. ... these can then be specified by the detector/geometry setup.

asalzburger commented on Core/include/Acts/Seeding/SeedfinderUtils.hpp:
```diff
@@ -0,0 +1,240 @@
+) {
+  if (s.radius() > 120) {
+    return false;
+  }
```
Magic number have to go ...

asalzburger commented on Core/include/Acts/Seeding/SeedfinderUtils.hpp:
```diff
@@ -0,0 +1,240 @@
+  double az = (res[1].max() / low.radius()) * std::max(std::abs(low.z() - config.collisionRegionMin), std::abs(low.z() - config.collisionRegionMax));
+  double p2 = 0.0005 * az;
+  double p3 = 0.0001 * std::sqrt(az);
+  double delta_phi = std::min(0.085, std::min(p2, p3));
```
Also here, we need to find a way to make those cuts configurable.

asalzburger commented on Core/include/Acts/Utilities/KDRange.hpp:
```diff
@@ -0,0 +1,430 @@
+///
+/// @tparam Type The scalar type of the values contained in this range.
+template <typename Type>
+class OneDRange {
```
I see, this could actually be useful also in the Geometry parsing ... 

... we called the `KDBinnedArray`  actually `BinnedArrayXD` - so for consistency, one could do the same here and call this `RangeXD`, while the actual classes could be called `Range1D`, `Range2D` ... etc.

asalzburger commented on Examples/Io/Performance/ActsExamples/Io/Performance/SeedingPerformanceWriter.cpp:
```diff
@@ -53,6 +55,11 @@ ActsExamples::SeedingPerformanceWriter::SeedingPerformanceWriter(
+  std::ofstream myfile;
+  myfile.open("eff.csv");
+  myfile << "event,phi,eta,pT,matched\n";
+  myfile.close();
```
We should avoid having ad-hoc files written ..,

asalzburger reviewed (commented):
This is my first set of comments, I did not have time yet to read through the algorithm body.

stephenswat commented:
Thank you Andy for taking a look and for your comments! I'll make sure to resolve the points you raised before we move to potentially merging this; they were mostly designed to help me debug things and you're right that I should clean them up.

stephenswat commented:
As discussed, I have split up this pull request, creating #985, #986, and #987. Hopefully that will make review easier. :smile:

asalzburger commented:
Hi, shall we make a push and try to integrate this into the Core? 

This is so much work + already well digested stuff, would be great to get it in.

stephenswat commented:
@asalzburger Okay, it's all been moved over to its own tool now, away from the examples directory. There are a few minor things to do still, but it's very close to ready now. Need #1158, #1159, and #1160 for this to work, but they are harmless. We can chat during the morning meeting tomorrow.

asalzburger commented:
Excellent, next on my list of feature reviews.

robertlangenberg commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.hpp:
```diff
@@ -0,0 +1,79 @@
+
+#pragma once
+
+#include "Acts/Geometry/Extent.hpp"
```
are you using this?

stephenswat replied:
Removed.

robertlangenberg commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.hpp:
```diff
@@ -0,0 +1,79 @@
+  tree_t createTree(
+      const std::vector<const internal_sp_t *> &spacePoints) const;
+
+  template <typename OutputIt>
+  void helper(const internal_sp_t &middle,
+              const std::vector<const internal_sp_t *> &bottom,
+              const std::vector<const internal_sp_t *> &top, OutputIt it) const;
+
+  template <std::size_t NDims, typename output_it_t>
+  void processM(const tree_t &tree, output_it_t out_it,
+                const typename tree_t::pair_t &middle_p) const;
```
would prefer a less generic naming and/or comments what this does

stephenswat replied:
Renamed to `processFromMiddleSP`. Also, comments added.

robertlangenberg commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.ipp:
```diff
@@ -0,0 +1,489 @@
+}
+
+template <std::size_t N, typename internal_sp_t>
+bool validTupleOrtho(
```
where is this declared? why is it never called?

stephenswat replied:
Removed.

robertlangenberg commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.ipp:
```diff
@@ -0,0 +1,489 @@
+enum Dim { DimPhi = 0, DimR = 1, DimZ = 2 };
+
+template <std::size_t N, typename external_spacepoint_t>
+RangeXD<N, float> validTupleOrthoRange(
```
comment: returns a range reduced to the minimum defined by the config parameters

stephenswat replied:
Comment added.

robertlangenberg commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.ipp:
```diff
@@ -0,0 +1,489 @@
+
+template <typename external_spacepoint_t>
+template <std::size_t NDims, typename output_it_t>
+void SeedFinderOrthogonal<external_spacepoint_t>::processM(
```
please add a line of comment for each cut applied... or at least for the less obvious ones :)

stephenswat replied:
Added!

robertlangenberg commented on Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/SeedingOrthogonalAlgorithm.hpp:
```diff
@@ -0,0 +1,111 @@
+    float zMax = 2000.;
+    float maxSeedsPerSpM = 1;
+    float cotThetaMax = 7.40627;  // 2.7 eta
+    float sigmaScattering = 50;
```
I understand this is copy-paste, but it really irks me, make it 3 or 5 or something pleeeeze :D

stephenswat replied:
Done, set to 5. :smile:

robertlangenberg commented on Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/SeedingOrthogonalAlgorithm.hpp:
```diff
@@ -0,0 +1,111 @@
+  /// @param lvl is the logging level
+  SeedingOrthogonalAlgorithm(Config cfg, Acts::Logging::Level lvl);
+
+  std::optional<std::pair<double, double>> validTriple(
+      const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> &b,
+      const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> &m,
+      const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> &t) const;
+
+  template <typename OutputIt>
+  void helper(
+      const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> &middle,
+      std::vector<const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> *>
+          &,
+      std::vector<const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> *>
+          &,
+      OutputIt) const;
+
+  ActsExamples::SimSeedContainer run(
+      const std::vector<const Acts::InternalSpacePoint<SimSpacePoint> *>
+          &spacePoints) const;
+
+  template <std::size_t NDims>
+  Acts::KDTree<NDims,
+               const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> *,
+               Acts::ActsScalar, std::array, 4>
+  createTree(const std::vector<const Acts::InternalSpacePoint<SimSpacePoint> *>
+                 &spacePoints) const;
+
+  template <std::size_t NDims, typename tree_t>
+  void processM(const tree_t &tree, ActsExamples::SimSeedContainer &seeds,
+                const typename tree_t::pair_t &middle_p) const;
```
move declarations to class

stephenswat replied:
All moved.

robertlangenberg commented on Examples/Scripts/Python/seeding.py:
```diff
@@ -103,49 +103,86 @@ def runSeeding(trackingGeometry, field, outputDir, s=None):
         ),
     )
 
-    gridConfig = acts.SpacePointGridConfig(
```
create new separate orthogonalSeeding.py and keep seeding.py?

asalzburger replied:
No strong opinions here.

stephenswat replied:
Yeah, the executable now takes --orthogonal flag which uses the new strategy, if the flag is not provided it will use the old strategy. :slightly_smiling_face:

robertlangenberg reviewed (changes requested):
Please address the relatively minor comments and this can go in :)

stephenswat commented:
See you in 6 months when I'll have the changes done. :sunglasses: 

(Joking)

asalzburger commented:
Hi @stephenswat - there are only a few comments, let's get this in!!!

stephenswat commented:
It will be done by the end of the day.

stephenswat commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.hpp:
```diff
@@ -0,0 +1,79 @@
+      const SeedFinderOrthogonal<external_spacepoint_t> &) = delete;
+  //@}
+
+  template <typename input_container_t, typename output_container_t>
+  void createSeeds(const input_container_t &spacePoints,
+                   std::back_insert_iterator<output_container_t> out_it) const;
+
+  template <typename input_container_t>
+  std::vector<seed_t> createSeeds(const input_container_t &spacePoints) const;
```
Comments added.

stephenswat commented on Examples/Algorithms/TrackFinding/src/SeedingOrthogonalAlgorithm.cpp:
```diff
@@ -0,0 +1,129 @@
+
+  // calculation of scattering using the highland formula
+  // convert pT to p once theta angle is known
+  m_cfg.seedFinderConfig.highland =
```
Where can I find that?

stephenswat replied:
The only Highland implementation I can find is in Fatras, and I imagine we don't want to import that into Acts core?

asalzburger replied:
[[Pointwise material][https://github.com/acts-project/acts/blob/main/Core/src/Propagator/detail/PointwiseMaterialInteraction.cpp]]

Uses an evaluation of the multiple scattering component, not sure it is useful though.

stephenswat commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.ipp:
```diff
@@ -0,0 +1,489 @@
+RangeXD<N, float> validMidToLowOrthoRange(
+    const internal_sp_t &mid,
+    const SeedFinderOrthogonalConfig<ActsExamples::SimSpacePoint> &config) {
+  float pM = mid.phi();
```
I cannot use `ActsScalar` here, because some of the tools I depend on (i.e. the `SeedFilter`) are hardcoded to use floats. :frowning:

stephenswat replied:
I did manage to change a few things to `ActsScalar`, for example the k-d tree now uses it for its coordinates.

asalzburger replied:
Ok, I think we can do some cleanup campaign here later.

stephenswat commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.ipp:
```diff
@@ -0,0 +1,489 @@
+  res[DimR].shrinkMin(rM - config.deltaRMax);
+  res[DimR].shrinkMax(rM - config.deltaRMin);
+
+  float frac_r = res[DimR].min() / rM;
```
Renamed.

stephenswat commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.ipp:
```diff
@@ -0,0 +1,489 @@
+template <typename internal_sp_t>
+bool validTuple(const internal_sp_t &low, const internal_sp_t &high,
+                float colMin, float colMax, float cotThetaMax) {
+  float rL = low.radius();
```
Let's continue the discussion in the over conversation.

stephenswat commented on Core/include/Acts/Seeding/SeedFinderOrthogonal.ipp:
```diff
@@ -0,0 +1,489 @@
+#include <type_traits>
+
+namespace Acts {
+enum Dim { DimPhi = 0, DimR = 1, DimZ = 2 };
```
Done.

stephenswat commented on Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/SeedingOrthogonalAlgorithm.hpp:
```diff
@@ -0,0 +1,111 @@
+      const Acts::InternalSpacePoint<ActsExamples::SimSpacePoint> &t) const;
+
+  template <typename OutputIt>
+  void helper(
```
Renamed to `filterCandidates`.

asalzburger commented on Core/include/Acts/Seeding/SeedFinderOrthogonalConfig.hpp:
```diff
@@ -0,0 +1,110 @@
+  float bFieldInZ = 2.08 * Acts::UnitConstants::T;
+  // location of beam in x,y plane.
+  // used as offset for Space Points
+  Acts::Vector2 beamPos{0 * Acts::UnitConstants::mm,
```
One more comment we should follow up, the beam position actually should be contextual to the event, but again - something for later.

paulgessinger replied:
That's actually a good point. I believe the *conventional* seed finder currently also has this in the config. This should be passed per invocation at some point.

robertlangenberg reviewed (approved):
can go ahead!

stephenswat commented:
I've added some Python tests for this seeder. @paulgessinger, when you have time, could you take a look to make sure I've done sane things here?

robertlangenberg commented:
@stephenswat 
E       TypeError: runSeeding() got an unexpected keyword argument 'useOrthogonal'

stephenswat commented:
For the love of all that is holy will you please work now!!

paulgessinger reviewed (approved):
Reapproving to get this in.
